go/azidentity/azure_identity_credential_adapter.go

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+)
+
+//defaultScope defined the default Scopes
+const defaultScope = "https://management.azure.com/.default"
+
+//AzureIdentityCredentialAdapter is the adapter interface
+type AzureIdentityCredentialAdapter interface {
```

Comment:
This needs to adapt an `azcore.Credential` to an `autorest.Authorizer`, then other constructs can build on top of them.  I think it should look like this.

```go
// AzureIdentityCredentialAdapter is used to adapt an azcore.Credential to an autorest.Authorizer
type AzureIdentityCredentialAdapter interface {
	autorest.Authorizer
}

// NewAzureIdentityCredentialAdapter returns an implementation of AzureIdentityCredentialAdapter.
func NewAzureIdentityCredentialAdapter(credential azcore.Credential, options azcore.AuthenticationPolicyOptions) AzureIdentityCredentialAdapter {
	policy := credential.AuthenticationPolicy(options)
	return &policyAdapter{p: policy}
}

type policyAdapter struct {
	p azcore.Policy
}

// WithAuthorization implements the autorest.Authorizer interface for type policyAdapter.
func (ca *policyAdapter) WithAuthorization() autorest.PrepareDecorator {
	return func(p autorest.Preparer) autorest.Preparer {
		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
			r, err := p.Prepare(r)
			if err != nil {
				return r, err
			}
			_, err = ca.p.Do(r.Context(), &azcore.Request{Request: r})
			if errors.Is(err, azcore.ErrNoMorePolicies) {
				err = nil
			}
			return r, err
		})
	}
}
```

Reply:
You could probably even omit `AzureIdentityCredentialAdapter` and have `NewAzureIdentityCredentialAdapter()` return an `autorest.Authorizer`

Reply:
Agree, I have changed it according this, but if using AuthenticationPolicy(), Error handling goes wrong with wrong azure credential. So I think using gettoken() is better.

Reply:
We _must_ use the underlying authentication policy as it handles properly refreshing the credential when it's near expiration.  Can you please clarify the problem with error handling?

Reply:
Presumably you're referring to if the provided credentials are invalid.  In the case of an `azidentity.AuthenticationFailedError` error type we need to wrap it in a type that implements the `adal.TokenRefreshError` interface (see below).  This will prevent the retry policy from retrying on invalid credentials.

```go
// WithAuthorization implements the autorest.Authorizer interface for type policyAdapter.
func (ca *policyAdapter) WithAuthorization() autorest.PrepareDecorator {
	return func(p autorest.Preparer) autorest.Preparer {
		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
			r, err := p.Prepare(r)
			if err != nil {
				return r, err
			}
			_, err = ca.p.Do(r.Context(), &azcore.Request{Request: r})
			if errors.Is(err, azcore.ErrNoMorePolicies) {
				return r, nil
			}
			var afe *azidentity.AuthenticationFailedError
			if errors.As(err, &afe) {
				err = &tokenRefreshError{
					inner: afe,
				}
			}
			return r, err
		})
	}
}

// tokenRefreshError implements the adal.TokenRefreshError interface
type tokenRefreshError struct {
	inner error
}

func (t *tokenRefreshError) Error() string {
	return t.inner.Error()
}

func (t *tokenRefreshError) Response() *http.Response {
	return nil
}

func (t *tokenRefreshError) Unwrap() error {
	return t.inner
}
```